libportal/portal.c

```diff
@@ -273,3 +280,200 @@ xdp_portal_new (void)
 {
   return g_object_new (XDP_TYPE_PORTAL, NULL);
 }
+
+/**
+ * xdp_portal_running_under_flatpak:
+ * @error: return location for a #GError pointer
+ *
+ * If you are trying to determine whether to use portals, consider using
+ * xdp_portal_running_under_sandbox() instead.
+ *
+ * Returns: %TRUE if the current process is running under a Flatpak sandbox, or
+ *   %FALSE if either an error was encountered or the process is running
+ *   unsandboxed
+ */
+gboolean
+xdp_portal_running_under_flatpak (GError **error)
+{
+  static _Thread_local gboolean decided = FALSE;
+  static _Thread_local gboolean under_flatpak = FALSE;
+  g_autofree char *root_path = NULL;
+  int root_fd = -1;
+  int info_fd = -1;
+
+  if (decided)
+    return under_flatpak;
+
+  /* Note: the rest of this function is copied from xdg-desktop-portal */
+  root_path = g_strdup_printf ("/proc/%u/root", getpid ());
+  root_fd = openat (AT_FDCWD, root_path, O_RDONLY | O_NONBLOCK | O_DIRECTORY | O_CLOEXEC | O_NOCTTY);
+  if (root_fd == -1)
+    {
+      if (errno == EACCES)
+        {
+          struct statfs buf;
+
+          /* Access to the root dir isn't allowed. This can happen if the root is on a fuse
+           * filesystem, such as in a toolbox container. We will never have a fuse rootfs
+           * in the flatpak case, so in that case its safe to ignore this and
+           * continue to detect other types of apps.
+           */
+          if (statfs (root_path, &buf) == 0 &&
+              buf.f_type == 0x65735546) /* FUSE_SUPER_MAGIC */
+            goto end;
+        }
+
+      /* Otherwise, we should be able to open the root dir. Probably the app died and
+         we're failing due to /proc/$pid not existing. In that case fail instead
+         of treating this as privileged. */
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                   "Unable to open %s", root_path);
+      goto end;
+    }
+
+  info_fd = openat (root_fd, ".flatpak-info", O_RDONLY | O_CLOEXEC | O_NOCTTY);
```

Comment:
> As a side note, the /.flatpak-info mechanism for identifying a process as untrusted to a portal is perhaps not ideal. [...] Perhaps an area for improvement if anyone has ideas?

Since Flatpak 1.1.2 you can use `$FLATPAK_ID`.

https://github.com/flatpak/flatpak/releases/tag/1.1.2

Reply:
Good to know but I don't think we can assume Flatpak is at least version 1.1.2, and that is only a possibility when checking your own Flatpak'd status, not that of another process, so it's not a replacement for the `/.flatpak-info` mechanism.

Reply:
> Since Flatpak 1.1.2 you can use $FLATPAK_ID.

This won't work for WebKit's flatpak-like sandbox, which only creates .flatpak-info and needs to be supported as well.